ouroboros-network/src/Ouroboros/Network/Snocket.hs

```diff
@@ -129,3 +134,115 @@ data Snocket m fd addr = Snocket {
+
+type SocketSnocket = Snocket IO Socket SockAddr
+
+-- | Create a 'Snocket' for the given 'Socket.Family'. In the 'bind' method set
+-- 'Socket.ReuseAddr` and 'Socket.ReusePort'.
+--
+socketSnocket
+  :: AssociateWithIOCP
+  -- ^ associate the socket with I/O CP.  We use it when we connect (it can be
+  -- used multiple times on a socket) and when we accept a socket.
+  --
+  -- Though it could be used in `open`, but that is going to be used in
+  -- a bracket so it's better to keep it simple.
+  --
+  -> SocketSnocket
+socketSnocket iocp = Snocket {
+      getLocalAddr   = Socket.getSocketName
+    , getRemoteAddr  = Socket.getPeerName
+    , addrFamily     = socketAddrFamily
+    , open           = openSocket
+    , openToConnect  = \addr -> openSocket (socketAddrFamily addr)
+    , connect        = \s a -> do
+#if !defined(mingw32_HOST_OS)
+        Socket.connect s a
+#else
+        Win32.Async.connect s a
+#endif
+    , bind = \sd addr -> do
+        let SocketFamily fml = socketAddrFamily addr
+        when (fml == Socket.AF_INET ||
+              fml == Socket.AF_INET6) $ do
+          Socket.setSocketOption sd Socket.ReuseAddr 1
+#if !defined(mingw32_HOST_OS)
+          -- not supported on Windows 10
+          Socket.setSocketOption sd Socket.ReusePort 1
+#endif
+        when (fml == Socket.AF_INET6)
+          -- An AF_INET6 socket can be used to talk to both IPv4 and IPv6 end points, and
+          -- it is enabled by default on some systems. Disabled here since we run a separate
+          -- IPv4 server instance if configured to use IPv4.
+          $ Socket.setSocketOption sd Socket.IPv6Only 1
+
+        Socket.bind sd addr
+    , listen   = \s -> Socket.listen s 1
+    , accept   = berkeleyAccept iocp
+    , close    = Socket.close
+    , toBearer = Mx.socketAsMuxBearer
+    }
+  where
+    openSocket :: AddressFamily SockAddr -> IO Socket
+    openSocket (SocketFamily family_) = do
+      sd <- Socket.socket family_ Socket.Stream Socket.defaultProtocol
+      associateWithIOCP iocp (Right sd)
+        -- open is designed to be used in `bracket`, and thus it's called with
+        -- async exceptions masked.  The 'associteWithIOCP' is a blocking
+        -- operation and thus it may throw.
+        `catch` \(e :: IOException) -> do
+          Socket.close sd
+          throwIO e
+        `catch` \(SomeAsyncException _) -> do
+          Socket.close sd
+          throwIO e
+      return sd
+
+
+
+-- | Create a snocket for the given 'Socket.Family'.
+--
+rawSocketSnocket
+  :: AssociateWithIOCP
+  -> SocketSnocket
+rawSocketSnocket iocp = Snocket {
+      getLocalAddr  = Socket.getSocketName
+    , getRemoteAddr = Socket.getPeerName
+    , addrFamily    = socketAddrFamily
+    , connect       = \s a -> do
+#if !defined(mingw32_HOST_OS)
+        Socket.connect s a
+#else
+        Win32.Async.connect s a
+#endif
+    , bind          = \fd addr -> Socket.bind fd addr
```

Comment:
I'm not sure about the missing `setSocketOption` here. Are these "raw" snockets only to be used with ephemeral port numbers on with non-InitatorAndResponder mux applications?

Reply:
we use it in:
* local clients - which are using unix sockets - so no need for binding;
* and in one of the tests
I think it would be cleaner to move this inside `#if` where `clientSnocket = rawSocketSnocket` is set.  This way it will be cleaner that it is only for local clients over unix sockets.